pkg/controller/certificatesigningrequests/venafi/venafi.go

```diff
@@ -0,0 +1,203 @@
+			Complete()
+	})
+}
+
+func NewVenafi(ctx *controllerpkg.Context) *Venafi {
+	return &Venafi{
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		certClient:    ctx.Client.CertificatesV1().CertificateSigningRequests(),
+		recorder:      ctx.Recorder,
+		clientBuilder: venaficlient.New,
+	}
+}
+
+// Sign attempts to sign the given CertificateSigningRequest based on the
+// provided Venafi Issuer or ClusterIssuer. This function will update the resource
+// if signing was successful. Returns an error which, if not nil, should
+// trigger a retry.
+// Since this signer takes some time to sign the request, this controller will
+// set a "pick ID" annotation value that is used to fetch the latest state of
+// the request in subsequent re-syncs.
+func (v *Venafi) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	resourceNamespace := v.issuerOptions.ResourceNamespace(issuerObj)
+
+	client, err := v.clientBuilder(resourceNamespace, v.secretsLister, issuerObj)
+	if apierrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+		v.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		log.Error(err, message)
+		return nil
+	}
+
+	if err != nil {
+		message := fmt.Sprintf("Failed to initialise venafi client for signing: %s", err)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorVenafiInit", message)
+		log.Error(err, message)
+		return err
+	}
+
+	var customFields []venafiapi.CustomField
+	if annotation, exists := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiCustomFieldsAnnotationKey]; exists && annotation != "" {
+		err := json.Unmarshal([]byte(annotation), &customFields)
+		if err != nil {
+			message := fmt.Sprintf("Failed to parse %q annotation: %s", experimentalapi.CertificateSigningRequestVenafiCustomFieldsAnnotationKey, err)
+			v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorCustomFields", message)
+			util.CertificateSigningRequestSetFailed(csr, "ErrorCustomFields", message)
+			_, err = v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	duration, err := pki.DurationFromCertificateSigningRequest(csr)
+	if err != nil {
+		message := fmt.Sprintf("Failed to parse requested duration: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorParseDuration", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorParseDuration", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	pickupID := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiPickupIDAnnotationKey]
+
+	// check if the pickup ID annotation is there, if not set it up.
+	if len(pickupID) == 0 {
```

Comment:
suggestion: `pickupID` was unknown to me, I would suggest adding a comment about that.

```suggestion
	// The signing process with Venafi is slow. The "pickupID" allows us to track
	// the progress of the certificate signing. It is set as an annotation the first
	// time the Certificate is reconciled.
	pickupID := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiPickupIDAnnotationKey]
	if len(pickupID) == 0 {
```